Add support for Redis as a disk-based queue backend

Lukas0907 opened the pull request.

Lukas0907 committed:
Add support for Redis as a disk-based queue backend.

Lukas0907 committed:
Remove logging statements

Lukas0907 committed:
Make hostname, port, db and prefix configurable

Lukas0907 committed:
Randomly generate queue prefix

Lukas0907 committed:
Start redis on Travis CI

Lukas0907 committed:
Save name of the queue to 'redis.queue' file

Gallaecio commented on .travis.yml:
```diff
@@ -50,6 +50,8 @@ install:
         source "$HOME/virtualenvs/$PYPY_VERSION/bin/activate"
       fi
   - pip install -U tox twine wheel codecov
+services:
+  - redis
```
According to https://docs.travis-ci.com/user/database-setup/#redis the right value would be `redis-server`.

That said, we want the tests to be executable locally as well, so using this feature of Travis CI may not be the right approach.

Lukas0907 replied:
Oh right, thanks. 

> That said, we want the tests to be executable locally as well, so using this feature of Travis CI may not be the right approach.

I thought about relying on the user having Redis started before executing the tests. But as we discussed yesterday, the better approach is probably to start Redis ourselves (similar to the mockserver.py approach). So I will implement that!

Lukas0907 committed:
Use info.json file similar to queuelib and delete it if queue is empty

Lukas0907 committed:
Extend queue tests to Redis-based disk queue

Lukas0907 committed:
redis -> redis-server in .travis.yml

Lukas0907 committed:
temp

Lukas0907 committed:
redis server start/stop

Lukas0907 committed:
scheduler test: make SCHDULER_DISK_QUEUE setting dynamic.

Lukas0907 committed:
Take settings from crawler (spider might not be set for crawler)

Lukas0907 committed:
Test scheduler with Redis as well

Lukas0907 committed:
Remove redis-server from .travis.yml: redis-server started manually

Gallaecio commented:
We need to make the tests as skipped on systems where there is no `redis-server` installed. At the moment, the tests fail with `FileNotFoundError: [Errno 2] No such file or directory: 'redis-server'`.

We should also make sure `redis-server` is installed in Travis CI, to make sure these tests are run in Travis CI always. You might want to enable https://travis-ci.org/github/Lukas0907/scrapy to make sure this works one you update the `install` part of `.travis.yml` accordingly.

Also, `tests/requirements-py3.txt` should now include `redis`, so that `tox` installs that package.

Finally, do tests pass for you if you run all of them in a single `tox` command? (`tox -e py -- tests/test_scheduler.py tests/test_squeues.py`). The latter file tests fail for me. I wonder if we need to start the server with a different port each time.

Lukas0907 committed:
travis: Run CI on all branches

Lukas0907 committed:
Add redis as test requirement

Lukas0907 committed:
Skip tests that require redis-server if not installed

Lukas0907 committed:
fix flake8 warnings

Lukas0907 commented:
@Gallaecio Thanks for the review! I made the following changes according to your feedback:

* I added skip decorators so that tests which require a running redis-server are not executed in case redis-server is not available (cannot be started).
* I think that redis-server is automatically available in travis, it is just not started by default.
* I have added the requirement `redis` to requirements-py3.txt so that tox installs it properly.
* Regarding the ports problem: I also experienced that sometimes some tests fail because connecting to the redis port fails (it does seem to work most of the time, though). I think choosing a random port might be better/would fix the problem but it might also introduce a problem if the randomly chosen port is already in use by another service. Do you think I should still choose a port randomly or do you have a better solution mind?
* I fixed some small stuff to satisfy flake8.

whalebot-helmsman commented:
> We need to make the tests as skipped on systems where there is no redis-server installed. 

@Gallaecio is there visible sign of redis tests execution in such case? My concern is we break something and no tests are run.

As we talk on a meeting, we have an intention to put these queues into separate repo in future. May be we can drop this requirement?

Lukas0907 commented:
> @Gallaecio is there visible sign of redis tests execution in such case? My concern is we break something and no tests are run.

pytest marks skipped tests with an "x" instead of a ".", so in theory, it's visible but easily overlooked. Maybe a separate tox env could be used where running these tests is mandatory.

whalebot-helmsman commented:
> pytest marks skipped tests with an "x" instead of a ".", so in theory, it's visible but easily overlooked. 

We can close eyes on how much effort is required to check this. But even different tox environments of same commit have different amount of skipped tests, e.g. `pypy3` has 64 https://travis-ci.org/github/Lukas0907/scrapy/jobs/695992054#L597, `asyncio` has 52 https://travis-ci.org/github/Lukas0907/scrapy/jobs/695992058#L609

whalebot-helmsman commented:
I think, we should choose redis port randomly. E.g. you develop these message queues on server with production redis server. New instance fails to obtain a port and you work with production instance. Test spoils production data, production data spoils the tests.

whalebot-helmsman commented:
For obtaining random port we can use next snippet

```
from contextlib import closing
import socket
from subprocess import Popen, PIPE

def find_free_port():
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    # OS-specific mechanism to obtain random port
    s.bind(('', 0))
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    return s


# pass open port to redis
with closing(find_free_port()) as s:
    p = Popen(['redis-server', '--port', str(s.getsockname()[1])], stdout=PIPE)
    p.wait()
```

Lukas0907 commented:
Thank you for the snippet!

>`p.wait()`

Wouldn't this block the function because it waits for redis-server to terminate?

Lukas0907 committed:
Use random port for redis-server while testing.

Lukas0907 commented:
@whalebot-helmsman I adjusted your snippet for the tests in 5a42069. Seems to work fine. Thanks again!

whalebot-helmsman commented:
> Wouldn't this block the function because it waits for redis-server to terminate?

It was an example how to run redis server outside of context unit-tests

> I adjusted your snippet for the tests 

Looks good

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +97,93 @@ def _pickle_serialize(obj):
+        except ImportError:
+            raise NotConfigured('missing redis library')
+
+        # If called from from_crawler() method, self.crawler is set.
```
Why not override `@from_crawler` as well and create with `self.crawler.settings`.
This way, `settings` argument becomes required and it's properly set by the factory methods

Lukas0907 replied:
I actually wanted to do it this way first, however:

```python
PickleFifoRedisQueue = _scrapy_serialization_queue(
    PickleFifoRedisQueueNonRequest
)
```

The queue class that is used by the "end user" is `PickleFifoRedisQueue` which is a `ScrapyRequeuestQueue` that extends `PickleFifoRedisQueueNonRequest`. `ScrapyRequestQueue` defines the method `from_crawler()` as follows:

```python
        @classmethod
        def from_crawler(cls, crawler, key, *args, **kwargs):
            return cls(crawler, key)

        def __init__(self, crawler, key):
            self.crawler = crawler
            self.spider = crawler.spider
            super(ScrapyRequestQueue, self).__init__(key)
```

I'm not sure what would be the cleanest way to solve this. Maybe we could use `scrapy.utils.misc.create_instance()` in the `__init__`? In that case, it would use `from_settings` and `from_crawler` from the queue implementation class if they are available or just the plain old `__init__` if not.

ejulio replied:
Since you had to change to hold a reference to `crawler`, we could simplify it have a reference to `settings` as it's the true dependency.
Unless, we'll use crawler for something else later.
Also, we shouldn't get the settings as an argument in `__init__`, as it is not set by the serialization queue, unless we change that interface as well.

Lukas0907 replied:
I tried to address your change requests in https://github.com/Lukas0907/scrapy/pull/1/commits/920d987a61c755ea6617e7211a3b7008e44f86a3. 

A `SerializableQueue` now optionally takes a `settings` argument. This should be backwards-compatible and allow new queue classes that require access to `settings`.

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +97,93 @@ def _pickle_serialize(obj):
+        logger.debug("Using redis queue '%s'", self.info['queue_name'])
+
+    @classmethod
+    def from_settings(cls, settings, path):
```
I guess it would be better to with `from_crawler` as it is the method used in other queues.
Also, this is styling preference, since this method calls `__init__`, I'd prefer that it came before in the source (basically, I'd be reading from top to bottom, minimizing the back and forth). But this is only for coding style..

Lukas0907 replied:
I use `from_settings()` only in tests:

```python
    def queue(self):
        return PickleFifoRedisQueue.from_settings(self.get_settings(), self.qpath)
```

But I will change the order of the methods.

ejulio replied:
This is tricky then.
As `from_settings` is a standard approach to instantiate the class.
And, in tests we should be testing the user path, no shortcuts.
If you are using `from_settings` to avoid some extra code, then you could refactor the tests to reuse this code.
The idea is that, tests can be example of how we should use the class, so having a method only for tests isn't that good

Lukas0907 replied:
I removed the `from_settings()` and the test code uses the same code path as the production code now.

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +97,93 @@ def _pickle_serialize(obj):
+                info = json.load(f)
+        else:
+            if not prefix:
+                prefix = "scrapy-{}".format(random.randint(0, 2**32 - 1))
```
`2 ** 32`

I think this could be the setting default value, so you could do `settings.get('key', ...)`

Lukas0907 replied:
I could make a setting for this but I don't know how useful it would be because I would argue that the random value is chosen between `0` and `2**32-1` is merely an implementation detail so that the queue name is most likely unique.

Either way, I added a space.

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +97,93 @@ def _pickle_serialize(obj):
+            settings = self.crawler.settings
+
+        self.client = redis.Redis(
+            host=settings['SCHEDULER_EXTERNAL_QUEUE_REDIS_HOST'],
```
Not completely sure, but better to use `settings.get`, otherwise we can get a `KeyError`.
If we use `settings.get` and it is `None` then we can raise a specific exception with the missing setting and how to solve it

Lukas0907 replied:
Alright, I changed this and raise a `NotConfigured` exception.

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +97,93 @@ def _pickle_serialize(obj):
+                'queue': 'redis',
+                'queue_name': "{}-{}".format(prefix, self.path)
+            }
+        return info
```
If it's always going to have two keys `queue` and `queue_name`, then I'd set both as class attributes.
If they may change, or if there could be more stuff, then the dict should be fine

Lukas0907 replied:
You are right, it's probably better to set `self.queue.name`. If more data is needed, it can still be changed to a dict.

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +97,93 @@ def _pickle_serialize(obj):
+            }
+        return info
+
+    def _saveinfo(self, info):
```
maybe `_info_path`?

Lukas0907 replied:
I agree. I was following the interface of queuelib here (https://github.com/scrapy/queuelib/blob/master/queuelib/queue.py#L119)

ejulio commented on tests/test_squeues.py:
```diff
@@ -189,3 +198,37 @@ def test_serialize_request_recursive(self):
+
+    def setUp(self):
+        self.redis_server = RedisServer()
+        self.redis_server.__enter__()
```
maybe specific methods, like `open` and `close`?
Not sure on the styling, but I think it's a bit weird to call `__something__` methods in code, usually they are a proxy for some language thing, like `__len__` for `len()`

Lukas0907 replied:
Agreed, I changed it to `start()` and `stop()` because RedisServer is not used within an context manager anyway.

Lukas0907 committed:
Coding style changes

Lukas0907 committed:
Check if host, port and db are set and not None

Lukas0907 committed:
infopath -> info_path

Lukas0907 committed:
Change RedisServer.{__enter__,__exit__} to {start,stop}

Lukas0907 commented:
@ejulio Thanks for the review! I tried to incorporate your suggestions, maybe you want to have a look.

Lukas0907 committed:
Add pytest option --has-redis and only run Redis tests if given

Gallaecio commented on scrapy/utils/test.py:
```diff
@@ -126,3 +127,11 @@ def get_from_asyncio_queue(value):
+
+def get_free_port():
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    # OS-specific mechanism to obtain random port
```
The final implementation should be multiplatform. Maybe you could make this a `TODO`, as to not to forget.

Lukas0907 replied:
Which platforms should be supported? Linux, Mac, Windows?

According to the documentation, the socket module seems to be support by all three of them:

> This module provides access to the BSD socket interface. It is available on all modern Unix systems, Windows, MacOS, and probably additional platforms.

In particular, the socket method does not seem to be limited to non-Windows [1], but I didn't try it. Either way, I created a ticket (issue #3) so I don't forget about it.

[1] https://docs.python.org/3/library/socket.html#socket.socket

Gallaecio replied:
My comment was based on the code comment, “OS-specific”. Maybe I misunderstood what it meant?

Lukas0907 replied:
Right, maybe @whalebot-helmsman can shed some light on this (just copied it from his snippet).

Lukas0907 replied:
I just tried the snipped on Windows, so this should be platform compatible.

Gallaecio commented:
The test failures seem Travis-specific. You should restart those test jobs in Travis, or close and reopen this pull request to trigger all jobs to be restarted.

Lukas0907 commented:
> The test failures seem Travis-specific. You should restart those test jobs in Travis, or close and reopen this pull request to trigger all jobs to be restarted.

I restarted the Travis jobs and the checks have passed now.

Lukas0907 committed:
_saveinfo() -> _save_info()

Lukas0907 committed:
Support additional py.test arguments in CI via the SCRAPY_CI_ARGS env var

whalebot-helmsman commented on scrapy/squeues.py:
```diff
@@ -88,6 +97,99 @@ def _pickle_serialize(obj):
+                info = json.load(f)
+        else:
+            if not prefix:
+                prefix = "scrapy-{}".format(random.randint(0, 2 ** 32 - 1))
```
Is every instance of redis-queue will have it's own prefix as a result?

Lukas0907 replied:
As discussed, I changed this back to a static prefix. Therefore every queue will have the same (configured) prefix.

whalebot-helmsman commented on scrapy/squeues.py:
```diff
@@ -88,6 +97,99 @@ def _pickle_serialize(obj):
+        self.client.close()
+
+    def _loadinfo(self, prefix):
+        info_path = self._info_path()
```
Is every instance of redis-queue has an mirror file on a disk?

Lukas0907 replied:
Yes, so that the crawling can be picked up if it's interrupted.

Lukas0907 replied:
Actually it's not needed for every queue to have its own file on disk if we use static prefixes. I have therefore removed the info file in commit c04d1d2775cb53ca28cf950c03828acbc829b849!

whalebot-helmsman commented on tests/test_scheduler.py:
```diff
@@ -183,6 +203,11 @@ class TestSchedulerOnDisk(BaseSchedulerOnDiskTester, unittest.TestCase):
 
+@pytest.mark.redis
+class TestSchedulerRedis(BaseSchedulerRedisTester, unittest.TestCase):
+    priority_queue_cls = 'scrapy.pqueues.ScrapyPriorityQueue'
```
Can you also add tests for `priority_queue_cls = 'scrapy.pqueues.DownloaderAwarePriorityQueue'`

Lukas0907 replied:
Thanks, added!

Lukas0907 committed:
Refactor queue settings

Lukas0907 committed:
Set self.queue_name instead of self.info and expect SCHEDULER_EXTERNAL_QUEUE_REDIS_PREFIX to be configured.

Lukas0907 committed:
Add redis scheduler test for DownloaderAwarePriorityQueue

Lukas0907 committed:
get_free_port() works under Windows as well

Lukas0907 committed:
Remove unused import

Lukas0907 committed:
Make _RedisQueue class share one connection pool

ejulio commented on scrapy/squeues.py:
```diff
@@ -29,6 +36,10 @@ def _serializable_queue(queue_class, serialize, deserialize):
 
     class SerializableQueue(queue_class):
 
+        def __init__(self, path, settings=None, *args, **kwargs):
```
Maybe not default to `None`, as it's being "required"

Lukas0907 replied:
The problem is that it's only required for a special type of `SerializableQueue`, namely `PickleFifoRedisQueueNonRequest`. It's not required for `PickleFifoDiskQueueNonRequest`.

In theory we could make it required (and change the interface for `*NonRequest` queues) because `*NonRequest` queues are only used in tests. However, they are also exported so I'm not sure if we want to maintain compatibility here.

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +99,92 @@ def _pickle_serialize(obj):
+        if host is None or port is None or db is None or prefix is None:
+            raise NotConfigured(
+                "Please configure "
+                + "SCHEDULER_EXTERNAL_QUEUE_REDIS_HOST, "
```
No need for concatenation, if wrapped in `()``python concatenates a sequence of strings

Lukas0907 replied:
Right, I changed it.

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +99,92 @@ def _pickle_serialize(obj):
+        db = self.settings.get('SCHEDULER_EXTERNAL_QUEUE_REDIS_DB')
+        prefix = self.settings.get('SCHEDULER_EXTERNAL_QUEUE_REDIS_PREFIX')
+        if host is None or port is None or db is None or prefix is None:
+            raise NotConfigured(
```
I see no problems here, but think that a specific exception for each setting would be better.
That leads to a though in scrapy.
@Gallaecio , what do you think of having something like `settings.get('MY_SETTING', required=True)` and it would raise `NotConfigured` if the value is not set?
I see this pattern of getting the value, and if not set raising the exception everywhere, so we could try to help it a bit

Lukas0907 replied:
Adding a `required` argument could be a good idea!

Btw. I looked at the code and `settings['foo']` is basically the same as `settings.get('foo')` (the key syntax does not raise an `KeyError`):

```python
    def __getitem__(self, opt_name):
        if opt_name not in self:
            return None
        return self.attributes[opt_name].value
```

Gallaecio replied:
> @Gallaecio , what do you think of having something like `settings.get('MY_SETTING', required=True)` and it would raise `NotConfigured` if the value is not set?

[This is what I thought](https://github.com/scrapy/scrapy/pull/4493#issuecomment-625368256) :smile:

Lukas0907 replied:
Very cool. Is there any chance that https://github.com/scrapy/scrapy/pull/4493 can get merged? If yes, should I rebase this PR on scrapy#4493?

ejulio replied:
I don't think so @Lukas0907 .
For now, you can raise specific exceptions for each setting (maybe create a private method in your file).
Also, it could be worth mentioning this example in that PR/issue where it would simplify the flow

Lukas0907 replied:
Alright, I implemented my own function in 1b5a468cf9c73e8049904453fdc7a2349eab38a9.

ejulio reviewed (approved):
Hey @Lukas0907 . Great work here!
Left some minor comments.
The PR looks good to me, let's wait for @Gallaecio and @whalebot-helmsman reviews as well

Lukas0907 committed:
String cat

whalebot-helmsman commented on scrapy/squeues.py:
```diff
@@ -88,6 +99,92 @@ def _pickle_serialize(obj):
+        db = self.settings.get('SCHEDULER_EXTERNAL_QUEUE_REDIS_DB')
+        prefix = self.settings.get('SCHEDULER_EXTERNAL_QUEUE_REDIS_PREFIX')
+        if host is None or port is None or db is None or prefix is None:
+            raise NotConfigured(
```
We have a backup path for problems with disk queue https://github.com/Lukas0907/scrapy/blob/ec98dabfab60283303a9208ccd8177d9f995ba72/scrapy/core/scheduler.py#L93-L98 Can you check how it plays with current redis logic. My concern is all requests end up in memory in case of redis queue without all options set.

Lukas0907 replied:
@whalebot-helmsman At the moment `_dqpush()` only handles `ValueError` exceptions:

```python
try:
    self.dqs.push(request)
except ValueError as e:  # non serializable request
    ...
```

Therefore a rather ugly stack trace is produced if `SCHEDULER_DISK_QUEUE=scrapy.squeues.PickleLifoRedisQueue` is set but there are connection/configuration problems.

I could catch the exception in `_dqpush()`, log an error/warning, and return (similar to what is already done in case serialization is not possible). However, this would mean that the requests would indeed be queued in memory ...

Do you think that we should exit instead?

Lukas0907 committed:
Add _get_required_settings() to raise exception if setting not configured

Lukas0907 committed:
Do not create a file for every redis queue

queue name is static and only depends on static prefix and path

Lukas0907 committed:
Make redis tests more robust by waiting for redis-server to accept connections

Lukas0907 committed:
Add docs for Redis

Lukas0907 committed:
In case there is a connection error, assume the queue is empty.

This allows a clean shutdown of Scrapy if there is a connection problem.

Lukas0907 committed:
Deal with errors when pushing to disk queues

Lukas0907 committed:
Deal with error when creating priority queue (which creates disk queue)

Lukas0907 committed:
add test for missing setting

whalebot-helmsman commented on docs/topics/jobs.rst:
```diff
@@ -82,3 +84,32 @@ running :class:`~scrapy.spiders.Spider` class.
+
+If the scheduler should store requests in Redis instead of on disk, the
+:setting:`SCHEDULER_DISK_QUEUE` setting has to be set to
+``scrapy.squeues.PickleLifoDiskQueue`` or
```
I think, here is a wrong queue name.

Lukas0907 replied:
Right, of course. Thanks!

Lukas0907 committed:
Fix doc

whalebot-helmsman commented on scrapy/core/scheduler.py:
```diff
@@ -90,7 +91,14 @@ def enqueue_request(self, request):
+        except Exception as e:
+            logger.error("Unable to push to disk queue: %(reason)s",
+                         {'reason': e})
+            return False
```
I think we need something like `dqok = False` here. E.g. we have a temooral connection to redis issue, so it is good to store requests in memory until issue resolved.

Lukas0907 replied:
Okay, that makes things even easier. This would also mean, however, that if the user messed up the Redis configuration from the start, Scrapy would just use a memory queue and issue a warning (error?) but not exit. Is this the desired behavior?

whalebot-helmsman replied:
Before answer this question we should address https://github.com/Lukas0907/scrapy/pull/1/files#r447613132

whalebot-helmsman commented on scrapy/core/scheduler.py:
```diff
@@ -90,7 +91,14 @@ def enqueue_request(self, request):
             self.df.log(request, self.spider)
             return False
-        dqok = self._dqpush(request)
+
```
This is common comment to both new `try ... except` blocks. 
For redis queue idea is mimic current disk queue interface. Raising exceptions from methods is also a part of interface. I don't know how current disk queues handles errors. Do they raise exceptions?

Lukas0907 replied:
At the moment (in master), only serialization errors are handled (-> `ValueError` is raised) and the memory queue is used as a fallback (`dok = False`). In case of any other internal error (e.g. disk full, permission problem, etc.) an exception is raised that is not handled by `enqueue_request()` and bubbles up the stack frames. Basically the same is true for the Redis implementation as well (it can also raise "internal exceptions").

The idea is now to handle these arbitrary exceptions in a way. One possibility is to not enqueue the request in such a case (effectively abort), this is how I solved it at the moment. Another way would be to fall back to the memory queue. I think both ways are fine in general.

Lukas0907 committed:
Fix log assert

whalebot-helmsman commented on docs/topics/settings.rst:
```diff
@@ -1207,7 +1207,55 @@ Default: ``'scrapy.squeues.PickleLifoDiskQueue'``
+
+.. setting:: SCHEDULER_EXTERNAL_QUEUE_REDIS_DB
+
+SCHEDULER_EXTERNAL_QUEUE_REDIS_DB
```
In `redis-py` library there is a way to pass all arguments using URL https://github.com/andymccurdy/redis-py/blob/a67a47d5c3f4cd22b4a9658aa4b182593906ec39/redis/client.py#L665-L694 . May be it is better to use it here instead of using 3 separate arguments?

Lukas0907 replied:
True and this would also allow the user to pass username and password and other settings. So I think this would be superior in every way!

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +98,67 @@ def _pickle_serialize(obj):
+        except ImportError:
+            raise NotConfigured('missing redis library')
+
+        host = self._get_required_setting('SCHEDULER_EXTERNAL_QUEUE_REDIS_HOST')
```
The documentation states some default values to these settings.
If they have default values, they aren't required :thinking: 
At least, they won't raise the expected exception..

To some extent, I think that only the host should be set, all other default values seem ok.

Lukas0907 replied:
> If they have default values, they aren't required :thinking: 
> At least, they won't raise the expected exception..

Yeah, that's true. The idea of the check is that if somebody deliberately sets the values to `None` ... but maybe it doesn't make sense to account for that case because we have the default values anyways?

ejulio replied:
It makes sense to have the check, better to fail with something we know.
Just didn't cross my mind the idea of someone removing the setting

ejulio commented on scrapy/squeues.py:
```diff
@@ -88,6 +98,67 @@ def _pickle_serialize(obj):
+    def _get_required_setting(self, name):
+        value = self.settings.get(name)
+        if value is None:
+            raise NotConfigured('Please configure {} in the project settings'
```
The message is good, but maybe we can be a bit more pragmatic :stuck_out_tongue: 
`"{} is not set in your project settings. For more information, please refer to the API Reference.."`
I'm completely sure, just thinking a bit differently about the message

cc @Gallaecio

Gallaecio replied:
I think we need to approach this with the minimum wording that would still help someone which is inheriting the code (so that person may not be even aware that Redis is used).

What about something in this lines:

```python
            raise NotConfigured(
                'When the SCHEDULER_DISK_QUEUE setting is defined as '
                '{queue}, the {name} setting must also be defined.'.format(
                    queue=repr(self.settings['SCHEDULER_DISK_QUEUE']),
                    name=name,
                )
            )
```

- We don’t mention the project settings, as there are multiple ways to define settings. Calling them “settings” should be enough to lead users to read the settings documentation.
- We explicitly mention the other setting which makes this setting mandatory. A user that is not aware of the custom value of `SCHEDULER_DISK_QUEUE` in the first place now can figure out where it is defined, and look it up in the documentation to understand that it does.

Lukas0907 replied:
@gallaecio Perfect, thank you!

Lukas0907 committed:
Use Redis URL instead of hostname + port + db

Lukas0907 committed:
Change wording for required but not configured settings

Gallaecio commented on scrapy/squeues.py:
```diff
@@ -88,6 +98,65 @@ def _pickle_serialize(obj):
+        if self.client is None:
+            # Note: We set the instance variable here.
+            # All RedisQueue objects share the same client object.
+            _RedisQueue.client = redis.Redis.from_url(url)
```
Is the queue able to work if the setting is defined in `Spider.custom_settings`? If so, I wonder how problematic this could be if [running two spiders at once](https://docs.scrapy.org/en/latest/topics/practices.html#running-multiple-spiders-in-the-same-process), each with a different Redis URL.

Lukas0907 replied:
Good point. I will investigate this.

Lukas0907 replied:
I added code to handle different URLs in 0cbdbfb4d4e1691f788c8553aba2784f570d2fdb. A client is now identified by the URL and only if the URL is the same, it will be reused.

I tried scraping with two spiders and custom settings that specify different redis URLs which worked fine.

Gallaecio reviewed (commented):
Great work on the documentation. I’ll have some structural/organizational feedback, but given that the goal is to move this code to a repository of its own instead of merging the changes into Scrapy, I think most of the documentation feedback can wait until then.

Lukas0907 committed:
Change exception message if one of the settings is None

Lukas0907 committed:
Reuse client only for the same URL

Lukas0907 committed:
pqueues: Run a self-check upon creation of the priority queue

Lukas0907 committed:
Fall back to memory queue if failing to push to disk queue

Lukas0907 committed:
Deal with all errors that can happen during queue creation

Lukas0907 committed:
Use random priority so that in case of LIFO an existing queue is not modified

Lukas0907 committed:
flake8 fix

Lukas0907 committed:
Find an empty/unused queue and check if it is really empty

Lukas0907 committed:
point out that the queue has to be deleted if the jobdir is deleted as well

Lukas0907 committed:
Add queue size to debug log

Lukas0907 committed:
Let scrapy fail if disk queue cannot be created and replace queue self-check with ping in Redis

Lukas0907 committed:
If prefix is missing, fail before doing a ping

ejulio commented on scrapy/squeues.py:
```diff
@@ -99,7 +99,7 @@ def _pickle_serialize(obj):
 
 class _RedisQueue(ABC):
-    client = None
+    clients = {}
```
I don't know about others opinions, but when creating an empty `dict` I prefer `dict` over `{}`.
As `{}` is also a notation for `set`, it can be a bit confusing on what is the type we're using

Gallaecio replied:
I would not enforce otherwise in Scrapy, but personally I still prefer `{}`.

I do wonder what your position is on non-empty dict, as I think the confusion argument (which I consider valid, just not strong enough for me) could also hold true for those, and I know some people prefer to use `dict(foo='bar')` over `{'foo': 'bar'}`.

ejulio replied:
My thought is just that `{'foo': 'bar'}` (`dict`) is different than `{'foo', 'bar'}` (`set`).
But `{}` could be both `dict` or `set`, so going through a quick code look could generate a bit of confusion.
So, I try to write it as `dict()` which is explicit and is easier to spot..

ejulio commented on scrapy/pqueues.py:
```diff
@@ -58,8 +58,12 @@ def __init__(self, crawler, downstream_queue_cls, key, startprios=()):
         self.key = key
         self.queues = {}
         self.curprio = None
+        self.selfcheck()
```
Nothing against it, but from a user perspective this could be a problem.
When I instantiate a class, I don't expect it to fail (raise an exception) from a side effect.
For example, when I create a database connection, usually the error is raised when the connection is `Open`, but not when it is created.
The reason I see for this kind of behavior is that we don't expect constructors to raise exceptions or have side effects (usually they should perform some local computation and keep an initial valid state in the class).

This is mainly an advice then a change request, but if you see that it would work almost the same by putting this check in `init_prios` for example, then I'd be more in favor there.

Another place that could be a candidate is `from_crawler`.
Even though, it's a kind of factory method and it should only create the object, this kind of check could go there, as an aternative.

Lukas0907 replied:
I totally agree, I'm also not really in favor of side effects in a constructor. I implemented it this way because the constructor also calls `init_prios()` which is also not side-effect free. (I can also move it to `init_prios()` but that would not change anything in regards to the side-effects of the constructor.)

Gallaecio replied:
I see this as a `ValueError`. And I think it makes sense for `__init__` methods to raise a `ValueError` if they get an invalid parameter value.

whalebot-helmsman commented on tests/test_scheduler.py:
```diff
@@ -175,6 +182,20 @@ def test_dequeue_priorities(self):
                          sorted([x[1] for x in _PRIORITIES], key=lambda x: -x))
 
 
+class BaseSchedulerRedisTester(BaseSchedulerOnDiskTester):
```
I am constantly forgetting about this question on our meetings. Which prefix do we use here, after we decided to not generate one randomly? Should we clear queues in some way here?

Lukas0907 replied:
Good question, this actually uses the default prefix but the jobdir is: `self.jobdir = tempfile.mkdtemp()` The queue name is based on the jobdir as well, so there no collisions are basically impossible.

However, clearing the queue is definitely a good idea. I will add this to the tearDown.

Lukas0907 replied:
I checked the code and tests again. They don't actually leave a queue behind because all items that are pushed are also popped. There is also a cleanup test. So I don't think that an explicit clearing of the queue is necessary.

whalebot-helmsman replied:
Yes, it is not a problem as we set up special redis instance for every test case and tear down it after

Gallaecio commented on scrapy/pqueues.py:
```diff
@@ -157,11 +161,15 @@ def __init__(self, crawler, downstream_queue_cls, key, slot_startprios=()):
             self.pqueues[slot] = self.pqfactory(slot, startprios)
 
+    def selfcheck(self):
+        queue = self.pqfactory('hostname.invalid')
+        queue.close()
```
:lipstick: In line with the equivalent from the previous queue:

```suggestion
        self.pqfactory('hostname.invalid').close()
```

Lukas0907 replied:
Fixed, thanks!

Gallaecio commented on tests/mockserver.py:
```diff
@@ -19,7 +19,7 @@
 from scrapy.utils.python import to_bytes, to_unicode
 from scrapy.utils.ssl import SSL_OP_NO_TLSv1_3
-from scrapy.utils.test import get_testenv
+from scrapy.utils.test import get_testenv, get_free_port
```
:lipstick: 

```suggestion
from scrapy.utils.test import get_free_port, get_testenv
```

Lukas0907 replied:
Fixed, thanks!

Gallaecio commented on tests/requirements-py3.txt:
```diff
@@ -9,6 +9,7 @@ pytest-twisted >= 1.11
 pytest-xdist
 sybil
 testfixtures
+redis
```
:lipstick: 

```suggestion
redis
sybil
testfixtures
```

Lukas0907 replied:
Fixed, thanks!

Gallaecio commented on tests/test_squeues.py:
```diff
@@ -1,17 +1,30 @@
+from scrapy.settings import Settings
+from tests.mockserver import RedisServer
+
+import logging
```
:lipstick:  This import should go to the first block of imports.

Lukas0907 replied:
Fixed, thanks!

Lukas0907 committed:
don't use temp variable

Lukas0907 committed:
reorder imports

Lukas0907 committed:
cleanup test: check that queue is empty

Lukas0907 committed:
ignore missing types in scrapy.squeues

Lukas0907 committed:
add queues interface doc

ejulio commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,68 @@
+If you want to use your own disk queue, it has to conform to the following
+interface:
+
+.. class:: MyExternalQueue
```
This is only the case if the user calls `_scrapy_serialization_queue`, right?
Otherwise, they need to define `from_crawler` or similar..

Maybe you should describe the full interface and mention scrapy provides these helper methods

Lukas0907 replied:
That's true, hm, I should make the difference and the Scrapy provided methods clearer.

Lukas0907 replied:
I changed the docs now so that they actually document the interface that is exposed through the `_scrapy_serialization_queue()` method/class.

ejulio commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,68 @@
+      number of elements.
+
+      If the push fails because of a temporary problem (e.g. the connection
+      was dropped), a ``TransientError`` should be raised. This causes the
```
What if it is not a transient error?
What defines a temporary error? Something that happens up to 5 minutes, 1 hour? The end of the process?

Lukas0907 replied:
I'm not really sure about that. What do you say @whalebot-helmsman?

whalebot-helmsman replied:
Permanent error will be never resolved during crawl. Such errors make external storage unusable and all requests will be stored in memory. Incorrect configuration is an example of permanent error. 
Transient is negation of permanent error. There is a possibility  of it's resolution during a crawl. As a result it has sense to use memory to store requests until it resolves. Examples of transient errors are network partition between crawler and message queue and unserializable request.

ejulio replied:
Makes sense, though I still have a "concern" on something transient that is not resolved during the crawl.
It doesn't need to be resolved here, but we should, at least, agree on a convention

ejulio commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,68 @@
+      pushed value (LIFO) or the least recently pushed value (FIFO) is
+      returned.
+
+      .. note::
```
I know this is the standard way in scrapy for now, but for consistency, I think it would be better to raise the error here as well.
The main reason is that, we can subclass the exceptions and have proper handlers for each subtype, if required..

Lukas0907 replied:
Sounds good to me.  What do you say @whalebot-helmsman?

whalebot-helmsman replied:
@ejulio do you have in mind special reaction from callers side? Otherwise adding exception here is only additional complication of interface without clear advantage.

ejulio replied:
At first, consistency.
I'm interacting with a remote queue.
When pushing, it gives me an error, but when popping it returns None?

Then, we could have some sort of information on the error as well.
For example, the error could have a wait time before being retried, instead of looping forever.

ejulio commented on docs/index.rst:
```diff
@@ -252,6 +253,9 @@ Extending Scrapy
     Quickly export your scraped items to a file (XML, CSV, etc).
 
+:doc:`topics/queues`
+    Store requests in a queue to pause and resume crawls.
```
I think _it stores requests in queues_, but to pause/resume they need to be disk queues

Lukas0907 replied:
Yeah maybe I should just leave "to pause and resume crawls" out.

ejulio reviewed (commented):
Maybe it's worth checking this idea
https://rhodesmill.org/brandon/2012/one-sentence-per-line/

whalebot-helmsman commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,68 @@
+      the currently active settings.
+
+      The constructor is expected to verify the arguments and the relevant
+      settings and raise a ``ValueError`` in case of an error. This may
```
Why is  `ValueError` so different from other exceptions?

Lukas0907 replied:
It actually does not matter because the exception is not handled anyway in `scheduler.py/_dq()`. I thought ValueError makes sense (and IIRC @Gallaecio suggested it in our meeting).

whalebot-helmsman replied:
In that case let's not specify exact type here. If I understand it correctly, error in constructor leads to crawl halt. If so let's add it ot the doc.

whalebot-helmsman replied:
May be we need some common statement about unreadiness of scrapy to handle exceptions from all these methods and how such exceptions leads to spider halt, if the opposite not stated. After that we can add a remark for `TransientError` raised from `push` method.

whalebot-helmsman commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,68 @@
+
+   .. method:: __len__(self)
+
+      Returns the number of elements in the queue. If the number of elements
```
Here and in other places we require caching of queue length. It is the way we discussed it during the meeting and the way we want to implement it here. But I see this documentation as usable outside of our GSoC project. Can we exclude part about caching and provide more generic way here?

Lukas0907 replied:
Should I rephrase it to something like that the number of elements in the queue must be returned even in case of connection problems, etc.?

whalebot-helmsman replied:
I was thinking about even something more generic. Something like - don't return 0 from this method if not sure.

whalebot-helmsman replied:
Otherwise scrapy will close this queue and won't dequeue requests from it.

whalebot-helmsman commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,68 @@
+======
+
+Scrapy uses queues to schedule requests. By default, a memory-based queue is
+used but if :ref:`crawls should be paused and resumed <topics-jobs>`, using an
```
Another use-case is storing more requests than can be stored in memory.

Lukas0907 replied:
Good point, I'll add it.

Lukas0907 committed:
change doc of queue api to actual external interface

Lukas0907 committed:
Refactor redis queue to conform to queue interface specification

Lukas0907 committed:
scheduler: Differentiate between non-serializable and transient error

whalebot-helmsman commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,88 @@
+         In case an exception is raised, the crawling process is halted.
+
+      :raises Exception: If ``key`` or a queue-specific setting is invalid.
+          Exceptions are not handled by the caller and the crawling process is
```
There is a duplication between this section and note before

Lukas0907 replied:
True, I will remove it in the raises block.

whalebot-helmsman commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,88 @@
+
+      If the push fails because of a temporary problem (e.g. the connection
+      was dropped), a ``TransientError`` should be raised. If the request
+      could not be serialized, a ``ValueError`` should be raised.
```
Let's have a special error instead of `ValueError` here.

Lukas0907 replied:
The `ValueError` is currently raised by:

```python
def _pickle_serialize(obj):
    try:
        return pickle.dumps(obj, protocol=4)
    # Both pickle.PicklingError and AttributeError can be raised by pickle.dump(s)
    # TypeError is raised from parsel.Selector
    except (pickle.PicklingError, AttributeError, TypeError) as e:
        raise ValueError(str(e)) from e
```

- Should I create a type, e.g. `SerializationError`, and change the function?
- Should the type inherit from `TransientError`?
- Should the type also inherit from `ValueError` (so that the change is backwards compatible)?

ejulio replied:
I think `SerializationError(ValueError)` should do the job.
I don't think serialization is transient, no matter how many I call it, I should get the same problem for the same object.

whalebot-helmsman commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,88 @@
+      used to convert the request to a dict which can then be easily
+      serialized with, for example, :meth:`pickle.dumps`.
+
+      If the push fails because of a temporary problem (e.g. the connection
```
This part is a duplication of `raises` section below.

Lukas0907 replied:
True, I consolidated it.

whalebot-helmsman commented on scrapy/pqueues.py:
```diff
@@ -58,8 +58,12 @@ def __init__(self, crawler, downstream_queue_cls, key, startprios=()):
         self.key = key
         self.queues = {}
         self.curprio = None
+        self.selfcheck()
```
Do we still need them as we create a separation between errors?

Lukas0907 replied:
Yes we need the selfcheck because otherwise `from_crawler()` does not raise an exception.

whalebot-helmsman replied:
As I understand you talk about case without `startprios`. In that case exception will be raised in first `push` method.  Is it different from what you want to achieve?

Lukas0907 replied:
Yes, exactly it's about having no `startprios`. The problem is that if we removed this, and the user mistyped e.g. the hostname, then this would result in a ConnectionError, hence in a TransientError and the memory queue would be used as a fallback.

Lukas0907 committed:
docs: DRY

whalebot-helmsman commented on docs/topics/queues.rst:
```diff
@@ -0,0 +1,83 @@
+
+      :raises TransientError: If pushing to the queue failed due to a
+          temporary error (e.g. the connection was dropped).
+      :raises ValueError: If pushing to the queue failed because the request
```
IIRC, we agreed to replace generic `ValueError` by more specific one.

Lukas0907 replied:
Yes, please see my comment here: https://github.com/Lukas0907/scrapy/pull/1#discussion_r458816687

whalebot-helmsman commented on scrapy/squeues.py:
```diff
@@ -88,6 +98,93 @@ def _pickle_serialize(obj):
+            string = self._pop()
+            self.size -= 1
+            return string
+        except redis.exceptions.ConnectionError as ex:
```
Do we want to handle https://github.com/andymccurdy/redis-py/blob/20ec5d771d5116f51735b1d7912447e75c8c3d53/redis/exceptions.py#L12 here?

Lukas0907 replied:
Handling `ConnectionError` and `TimeoutError` is probably a good idea because both of them could be considered as transient errors, e.g. here: https://github.com/andymccurdy/redis-py/blob/master/redis/client.py#L913

Lukas0907 committed:
Use SerializationError for failed serialization attempts

Lukas0907 committed:
Redis: Handle ConnectionError and TimeoutError
